### DISCUSSION TITLE
Add argument to digit-char to support different "alphabet"

### DISCUSSION THREAD

**Discussion by fstamour** [fstamour/cl-car#14] [Ideas]
It could be nice if digit-char could take a third argument for the alphabet. I could help to implement things like base64 encoding.

Current signature: `digit-char weight &optional radix => char`

https://www.lispworks.com/documentation/HyperSpec/Body/f_digit_.htm

**Comment by fstamour** [comment]
:thinking: If you pass an "alphabet" to digit-char, it should be able to infer the radix from it. So maybe it doesn't need a new parameter?

**Reply by Gleefre** [reply]
Currently `digit-char` (and `digit-char-p`) can be implemented more efficiently because of the layout of digits/letters A-Z in Unicode - they have consecutive codes, for example in SBCL `digits-char` is written like this:
```lisp
(defun digit-char (weight &optional (radix 10))
  "All arguments must be integers. Returns a character object that represents
a digit of the given weight in the specified radix. Returns NIL if no such
character exists."
  (and (typep weight 'fixnum)
       (>= weight 0) (< weight radix) (< weight 36)
       ;; Notice the "magic" numbers:
       ;;   48 = (char-code #\0)
       ;;   55 = (char-code #\A) - 10
       (code-char (if (< weight 10) (+ 48 weight) (+ 55 weight)))))
```

This makes passing the alphabet quite different from passing the radix.

**Comment by fstamour** [comment]
Using the new `digit-char` would not be backward-compatible. It might need to be exposed in a new package...

Then this could only be a library... still it would be nice.

**Comment by Gleefre** [comment]
It might be nice to make the same addition to the `digit-char-p`, but there is a potential problem - it is not as simple as the `digit-char` as multiple characters per digit are possible (for example `#xA` is the same as `#xa`; but there weirder examples like `६` is read as `6` on SBCL and ECL for example, since it is a decimal digit in Unicode - `#\DEVANAGARI_DIGIT_SIX` (sbcl's name).)

**Comment by Gleefre** [comment]
While it could only be a library, I think it would be cool if alphabets were integrated into the printer & reader - if it was possible to bind `*print-base*` and maybe `*read-base*` to an alphabet.

**Reply by fstamour** [reply]
interesting idea
